Alternate rand, rand_r and srand implementation

thejpster opened the pull request:
Builds on #31 but removes the thread-safety (as the standard doesn't require it) and supports both 16-bit `c_int` and 32-bit `c_int` (with RAND_MAX always being `c_int::MAX`).

Sympatron committed:
Implement rand_r

Sympatron committed:
Implement rand and srand

Sympatron committed:
Use rand_r algorithm for rand

Sympatron committed:
Export RAND_MAX symbol

Sympatron committed:
Require CAS for rand

Sympatron committed:
Small optimizations to rand

thejpster committed:
Update random number generator

thejpster committed:
Rand example also prints min/max

jonathanpallant committed:
Allow RAND_MAX to be set to 32767.

Might be useful if your newlib assumes that, even with 32-bit integers.

Also adds "rand" to list of features in "all" and fixes missing portable-atomic.

jonathanpallant committed:
Remove rand example.

It's not actually that useful.

Sympatron commented on src/rand_r.rs:
```diff
@@ -0,0 +1,68 @@
+/// Passing NULL (core::ptr::null()) gives undefined behaviour.
+#[cfg_attr(not(feature = "rand_r"), export_name = "tinyrlibc_rand_r")]
+#[cfg_attr(feature = "rand_r", no_mangle)]
+pub unsafe extern "C" fn rand_r(seedp: *mut c_ulong) -> c_int {
```
Why did you change it to `c_ulong`? I am pretty sure the signature is `int rand_r(unsigned int *seedp)`.

jonathanpallant replied:
Fair point (https://github.com/eblot/newlib/blob/2a63fa0fd26ffb6603f69d9e369e944fe449c246/newlib/libc/include/stdlib.h#L162).

I was worried that a 16-bit integer seed might not hold enough state. But perhaps it's fine. I'll test it and see.

jonathanpallant replied:
OK, I wrote a demo that forces the seed to be held as a 16-bit value, and the output still looks pretty well distributed.

<img width="1513" alt="image" src="https://github.com/user-attachments/assets/286328ff-7306-4f31-9a94-52bdf657258b">

jonathanpallant committed:
Use an 'unsigned int' as the rand_r seed.

This then matches the stdlib.h declaration.

Looking at the numbers generated, having that be a 16-bit value doesn't
appear to affect the quality of the output significantly. It's certainly
good enough for typical use cases.

jonathanpallant commented:
Is this one good to go?
